### PR TITLE
Allow MOTECH to create an OpenMRS patient if a case is not found

### PR DESCRIPTION
**Pull request by kaapstorm**
Feature flag: "Enable OpenMRS integration"

Extends MOTECH OpenMRS integration to create OpenMRS patients from CommCases cases.

The default behaviour is preserved (don't create patients).

Tests and documentation will follow in a later PR.

### CODE REVIEW

**Comment by kaapstorm** [comment]
Ended up just sticking tests and docs in here to avoid a new PR building on an unmerged PR.

This should be good to go, original code buddy @proteusvacuum and/or new code buddy @sravfeyn. I'm not sure how code review etiquette should work here, but if either of you could check this out, I'd appreciate it.

If you'd like a little more context than the description of the PR, try checking out the changes in the ["Docs" commit](https://github.com/dimagi/commcare-hq/pull/21542/commits/b411a3849eac08dc96338a441a3abda78ae54df9) before looking at the code, or just ping me.

File: corehq/motech/openmrs/finders.py [hunk 1]
```diff
@@ -54,6 +55,9 @@ class PatientFinder(DocumentSchema):
     Subclasses must implement the `find_patients()` method.
     """
 
+    # If no patients are found, should a new one be created?
```
Comment (sravfeyn) [inline corehq/motech/openmrs/finders.py]: I assume this is an explanation of the property below, but it does sound like a question to the reader

File: corehq/motech/openmrs/repeater_helpers.py [hunk 2]
```diff
@@ -554,14 +554,63 @@ def get_patient_id_type_uuids_values():
...
+            '/ws/rest/v1/patient/',
+            json=patient,
+        )
+        if 200 <= response.status_code < 300:
+            # response.json() is not the full patient record. We need
+            # the patient's identifiers and attributes.
+            return get_patient_by_uuid(requests, response.json()['uuid'])
+
+
+def find_patient(requests, domain, info, openmrs_config):
```
Comment (sravfeyn) [inline corehq/motech/openmrs/repeater_helpers.py]: Could rename to `find_or_create_patient` if that's what it does.

**Review by sravfeyn — APPROVED** [review]

**Review by sravfeyn — APPROVED** [review]

**Comment by kaapstorm** [comment]
Thanks @sravfeyn

File: corehq/motech/openmrs/repeater_helpers.py [hunk 3]
```diff
@@ -554,14 +554,63 @@ def get_patient_id_type_uuids_values():
...
+        if identifiers:
+            patient['identifiers'] = identifiers
+        response = requests.post(
+            '/ws/rest/v1/patient/',
+            json=patient,
+        )
+        if 200 <= response.status_code < 300:
+            # response.json() is not the full patient record. We need
+            # the patient's identifiers and attributes.
+            return get_patient_by_uuid(requests, response.json()['uuid'])
```
Comment (proteusvacuum) [inline corehq/motech/openmrs/repeater_helpers.py]: curious, why do you pass `requests` around? Isn't that an external library?

**Review by proteusvacuum — APPROVED** [review]

File: corehq/motech/openmrs/repeater_helpers.py (hunk 3 above)
Comment (kaapstorm) [reply corehq/motech/openmrs/repeater_helpers.py]: It's an instance of [corehq.motech.requests.Requests](https://github.com/dimagi/commcare-hq/blob/6a382796aa76521b21d5251ea914d5cb8b2ccfce/corehq/motech/requests.py#L45). It has some of the same methods as the library (`get`, `post`, and `delete`) but it does some extra stuff for you, like work out the full URL, authenticate the request, log the response, etc.

**Pull request merged**